working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,937 @@
+the type of `this` in the body of an extension type _E_: It is the on-type
+of _E_. Similarly, extension methods of _E_ invoked in the body of _E_ are
+subject to the same treatment as previously, which means that extension
+methods of the enclosing extension type can be invoked implicitly, and
+extension methods are given higher priority than instance methods on `this`
+when `this` is implicit.*
+
+
+### Prevent implicit invocations: Keyword 'type'
+
+This section specifies the effect of including the keyword `type` in the
+declaration of an extension type.
+
+*A reminder, to set the scene: An implicit extension member invocation is
+an invocation of the form `e.m()` (or similar for getter/setter/operator
+invocations) which invokes a member of an extension `E`, but where the
+static type of `e` is a type `T` that matches the on-type of `E`. This is
+the typical usage for the extension methods mechanism that Dart has had
+since version 2.6, and it is available with no changes for an extension
+declaration that does not include the keyword `type`. The other kinds of
+invocations of extension members are explicit extension member invocations
+(assuming that `E` is an extension type, an example would be
+`E(e).m()`) and typed extension member invocation (like `e.m()` where the
+static type of `e` is an extension type).*
+
+Let `E` be an extension declaration such that the keyword `extension` is
+followed by `type`. We say that `E` is an _explicit_ extension type
+declaration, and that it introduces an _explicit_ extension type.
+
+An explicit extension type declaration is not applicable for an implicit
+extension method invocation.
+
+*In other words, methods of an explicit extension type cannot be called on
+the on-type, only on the extension type. Otherwise, it works the same as an
+extension without the `type` modifier. For example:*
+
+```dart
+extension type Age on int {
+  Age get next => this + 1;
+}
+
+void main() {
+  int i = 42;
+  i.next; // Compile-time error.
+  Age age = 42;
+  age.next; // OK.
+}
+```
+
+*The terminology using the word 'explicit' is motivated by two things: (1)
+The extension type "is not implicit", because it doesn't allow for implicit
+extension member invocations, hence it "is explicit". (2) The extension
+type explicitly says `type`, and the extension type must be the static type
+of a receiver in order to invoke any members of the extension type, which
+is achieved by explicitly declaring it as a variable type, a return type,
+etc.*
+
+Let `E` be an explicit extension type declaration, and consider an
+occurrence of an identifier expression `id` in the body of an instance
+member of `E`. If a lexical lookup of `id` yields a declaration of a
+member of `E`, the expression is treated as `let v = this in v.id`
+where the static type of `v` is the enclosing extension type.
+A similar rule holds for function invocations of the form `id(args)`, and
+for operator invocations of the form `this OP arg` or `OP arg`.
```

Comment:
Is `this OP arg` a special carve out? or can I write `this.id` as well?

Reply:
I just deleted the reference to operators: We have so many locations in our specification documents where we skip over operators and assume that they can be treated just like a corresponding method (so `a + b` is treated as `a.plus(b)` where `plus` is assumed to correspond to the given `operator +`). That should work here, too.